### PR TITLE
Add SCTP support to agnhost connect / porter commands.

### PR DESCRIPTION
**Pull request by fedepaol**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:

This PR adds SCTP support to the `connect` and `porter` commands of the `agnhost` image used for network e2e tests. This is needed as those commands are used for network policy tests, and as part of https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md#sctp-connectivity-tests we want to extend coverage to SCTP too.

Originally this was started in https://github.com/kubernetes/kubernetes/pull/89582 , but splitting the PR makes sense, following @danwinship 's suggestion.
This also contains the changes suggested in https://github.com/kubernetes/kubernetes/pull/89582 .

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
[KEP] : https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0015-20180614-SCTP-support.md
```

### CODE REVIEW

**Comment by danwinship** [comment]
@fedepaol you need to `./hack/update-bazel.sh`
also, bump `test/images/agnhost/VERSION`

**Comment by fedepaol** [comment]
> @fedepaol you need to `./hack/update-bazel.sh`
> also, bump `test/images/agnhost/VERSION`

Done, also updated the depending images as I did in https://github.com/kubernetes/kubernetes/pull/89830

**Comment by danwinship** [comment]
sig-testing people: the existing NetworkPolicy tests use `nc` as a client and `agnhost porter` as the server, but the version of `nc` in `agnhost` only supports TCP and UDP, so that won't work for testing SCTP policies. So this adds SCTP support to `agnhost connect` and `agnhost porter` and then we'll update the NetworkPolicy tests to use `agnhost connect` instead of `nc`.

**Comment by claudiubelu** [comment]
I assume there are some tests which need this change, right? Which are those those tests?

**Comment by danwinship** [comment]
The new tests are in #89582

**Comment by danwinship** [comment]
/priority important-soon
ping

File: test/images/kitten/BASEIMAGE [hunk 1]
```diff
@@ -1,8 +1,8 @@
-linux/amd64=REGISTRY/agnhost:2.14-linux-amd64
-linux/arm=REGISTRY/agnhost:2.14-linux-arm
-linux/arm64=REGISTRY/agnhost:2.14-linux-arm64
-linux/ppc64le=REGISTRY/agnhost:2.14-linux-ppc64le
-linux/s390x=REGISTRY/agnhost:2.14-linux-s390x
-windows/amd64/1809=REGISTRY/agnhost:2.14-windows-amd64-1809
-windows/amd64/1903=REGISTRY/agnhost:2.14-windows-amd64-1903
-windows/amd64/1909=REGISTRY/agnhost:2.14-windows-amd64-1909
+linux/amd64=REGISTRY/agnhost:2.16-linux-amd64
```
Comment (danwinship) [inline test/images/kitten/BASEIMAGE:1]: @fedepaol I don't think you meant to make these changes in `test/images/kitten` and `test/images/nautilus`?

Comment (fedepaol) [reply test/images/kitten/BASEIMAGE:1]: It was intentional, but I missed updating the version to the one we are building.
Here https://github.com/kubernetes/kubernetes/issues/89782#issuecomment-608340939 I was pointed out to https://github.com/kubernetes/kubernetes/tree/master/test/images#making-changes-to-images which says
"Some test images (agnhost) are used as bases for other images (kitten, nautilus). If the parent image's VERSION has been bumped, also bump the version in the children's BASEIMAGE files in order for base image changes to be reflected in the child images as well."

Now, it looks like this is not being done so often (I rebased the agnost version but did not have to rebase these files), so either removing these or bumping up to the updated version is fine to me.

Also pinging @claudiubelu as he gave me the instruction link last time.

Comment (claudiubelu) [reply test/images/kitten/BASEIMAGE:1]: Yeah, in fact, those values will have to be bumped to 2.20 (the version this PR bumps agnhost to). This is especially needed by the Image Builder / Promoter job, as it was out of order for a while, and there's no agnhost.2.16 in staging. Same about the nautilus image.

File: test/images/agnhost/porter/porter.go [hunk 2]
```diff
@@ -29,23 +29,24 @@ import (
...
 var CmdPorter = &cobra.Command{
 	Use:   "porter",
 	Short: "Serves requested data on ports specified in ENV variables",
-	Long: `Serves requested data on ports specified in ENV variables. For example, if the environment variable "SERVE_PORT_9001" is set, then the subcommand will start serving on the port 9001.
-
-Additionally, if the environment variable "SERVE_TLS_PORT_9002" is set, then the subcommand will start a TLS server on that port.
-
-The included "localhost.crt" is a PEM-encoded TLS cert with SAN IPs "127.0.0.1" and "[::1]", expiring in January 2084, generated from "src/crypto/tls".
-
-To use a different cert/key, mount them into the pod and set the "CERT_FILE" and "KEY_FILE" environment variables to the desired paths.`,
```
Comment (claudiubelu) [inline test/images/agnhost/porter/porter.go:48]: The information about the certificate / key is still useful and should still be kept, IMO.

**Review by claudiubelu — COMMENTED** [review]
Looks fine.  I've enabled the SCTP feature on my 1.19 cluster and I've tried running those tests you've mentioned on a Linux node with these changes (built them into ``claudiubelu/agnhost:2.13``), but they seem to be failing for me. Any ideas?

Logs: https://paste.ubuntu.com/p/Gw2BPmbWvh/

File: test/images/agnhost/porter/porter.go (hunk 2 above)
Comment (fedepaol) [reply test/images/agnhost/porter/porter.go:48]: Agree

**Comment by fedepaol** [comment]
> Looks fine. I've enabled the SCTP feature on my 1.19 cluster and I've tried running those tests you've mentioned on a Linux node with these changes (built them into `claudiubelu/agnhost:2.13`), but they seem to be failing for me. Any ideas?
> 
> Logs: https://paste.ubuntu.com/p/Gw2BPmbWvh/

Gave a look at the logs. 
The pods are connecting, meaning that the image is working properly:
```
STEP: Testing pods can connect to both ports when no policy is present.
STEP: Creating client pod client-can-connect-80 that should successfully connect to svc-server.
Jun 18 14:08:28.403: INFO: Waiting for client-can-connect-80-7zzzj to complete.
Jun 18 14:08:34.798: INFO: Waiting for client-can-connect-80-7zzzj to complete.
Jun 18 14:08:34.798: INFO: Waiting up to 5m0s for pod "client-can-connect-80-7zzzj" in namespace "sctp-network-policy-3964" to be "Succeeded or Failed"
Jun 18 14:08:35.010: INFO: Pod "client-can-connect-80-7zzzj": Phase="Succeeded", Reason="", readiness=false. Elapsed: 211.879808ms
STEP: Saw pod success
Jun 18 14:08:35.010: INFO: Pod "client-can-connect-80-7zzzj" satisfied condition "Succeeded or Failed"
STEP: Cleaning up the pod client-can-connect-80-7zzzj
STEP: Creating client pod client-can-connect-81 that should successfully connect to svc-server.
Jun 18 14:08:35.959: INFO: Waiting for client-can-connect-81-9fwqw to complete.
Jun 18 14:08:40.352: INFO: Waiting for client-can-connect-81-9fwqw to complete.
Jun 18 14:08:40.352: INFO: Waiting up to 5m0s for pod "client-can-connect-81-9fwqw" in namespace "sctp-network-policy-3964" to be "Succeeded or Failed"
Jun 18 14:08:40.549: INFO: Pod "client-can-connect-81-9fwqw": Phase="Succeeded", Reason="", readiness=false. Elapsed: 196.250154ms
STEP: Saw pod success
Jun 18 14:08:40.549: INFO: Pod "client-can-connect-81-9fwqw" satisfied condition "Succeeded or Failed"
STEP: Cleaning up the pod client-can-connect-81-9fwqw
```

What's not working, is having the policies blocking the connections:
```
Jun 18 14:08:47.097: FAIL: Pod client-cannot-connect-2vtkf should not be able to connect to service svc-server, but was able to connect.
```

This may mean either the tests are broken (but I tested them against an OCP where I know sctp is supported by the sdn) or that the sdn you are using does not support sctp policies.

**Comment by danwinship** [comment]
> This may mean either the tests are broken (but I tested them against an OCP where I know sctp is supported by the sdn) or that the sdn you are using does not support sctp policies.

I see in the logs:

    Jun 18 14:07:54.787: INFO: 1 / 1 pods ready in namespace 'kube-system' in daemonset 'azure-cni-networkmonitor' (0 seconds elapsed)

And googling turns up https://github.com/Azure/azure-container-networking as the likely source of that. There are no explicit references to SCTP in the source, but also there doesn't seem to be any obvious attempt to limit NetworkPolicy to just TCP and UDP...

**Comment by fedepaol** [comment]
> > This may mean either the tests are broken (but I tested them against an OCP where I know sctp is supported by the sdn) or that the sdn you are using does not support sctp policies.
> 
> I see in the logs:
> 
> ```
> Jun 18 14:07:54.787: INFO: 1 / 1 pods ready in namespace 'kube-system' in daemonset 'azure-cni-networkmonitor' (0 seconds elapsed)
> ```
> 
> And googling turns up https://github.com/Azure/azure-container-networking as the likely source of that. There are no explicit references to SCTP in the source, but also there doesn't seem to be any obvious attempt to limit NetworkPolicy to just TCP and UDP...

Interesting!
I just gave a quick look, I think the protocol is being handled here https://github.com/Azure/azure-container-networking/blob/5e26843be617ec0a597ebb5c80d79eb293180f64/npm/translatePolicy.go#L20 casting it to string and adding as -p to iptables. 
Will dig a bit more into it & retest it on OpenShift where I tested it before.
In any case, this should affect the tests themselves and not the image.

**Comment by danwinship** [comment]
> In any case, this should affect the tests themselves and not the image.

ah, true; the logs show that `agnhost porter` and `agnhost connect` are working as expected. If there is a bug, it's in #89582, not here

**Comment by claudiubelu** [comment]
I see. Then, I think we can get this in, and iron out the tests afterwards.

/lgtm

**Pull request merged**